### PR TITLE
[WIP] Display inherited Property Definitions

### PR DESCRIPTION
**Pull request by milesstoetzner**
The MR has the following changes for https://github.com/OpenTOSCA/winery/issues/238.

Note
- **the changes work for WPD**
- _IWineryRepositoryCommon#getParents has been introduced to receive all parents of an entityType_
- ConstraintClauseKV#equals function has been changed to check additional properties
- PropertyDefinitionKV#equals function has been changed to check additional properties

GET propertiesDefinition/merged
- returns properties definition considering inheritance

GET propertiesDefinition/inherited
- returns for each parent the propertiesDefinition (not considering inheritance)

POST propertiesDefinition
- store only definitions which are not defined by any parent

GET properties
- returns ONLY the stored values
- does not set default values anymore
- does not store properties back anymore
   
GET properties/effective
- returns the stored value or the default of each property

PAGE properties
-  in addition to self-defined properties display inherited properties
- GET propertiesDefinition/merged is used to render the form
- GET properties is used to fill the form
- form should not have any default values since on a POST the value would be set
- form is validated
    
PAGE properties defintions:
- display self-defined, overrriden and inherited properties definition in the same table with a corresponding status
- display self-defined properties definition of each parent in separate table

### CODE REVIEW

File: org.eclipse.winery.frontends/app/tosca-management/src/app/instance/sharedComponents/properties/properties.types.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ * which is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+ *******************************************************************************/
+
+export interface Properties {
+    [key: string]: string;
+}
+
+export interface PropertiesData {
```
Comment (milesstoetzner) [inline org.eclipse.winery.frontends/app/tosca-management/src/app/instance/sharedComponents/properties/properties.types.ts:19]: is this still valid considering a response in XML?

File: org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/entitytemplates/PropertiesResource.java [hunk 2]
```diff
@@ -75,47 +74,62 @@ public Response setProperties(LinkedHashMap<String, String> properties) {
...
-        // no Winery special treatment, just return the XML properties
-        // These can be null resulting in 200 No Content at the caller
-        if (props == null) {
-            return Response.ok().entity("{}").type(MediaType.APPLICATION_JSON).build();
+        TEntityTemplate.Properties properties = this.template.getProperties();
+        WinerysPropertiesDefinition propertiesDefinition = RepositoryFactory.getRepository()
+            .getTypeForTemplate(this.template)
+            .getWinerysPropertiesDefinition();
+
+        // TODO: apply merge logic to all props types
```
Comment (milesstoetzner) [inline org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/entitytemplates/PropertiesResource.java:96]: We need to discuss this.

Properites are not updated when properties definitions changes. 
In general, this is not a problem, but the frontend renders a form based on the returned properties, and, therefore, does not render new definitions.

Therefore, reconstruct default properties all the time and fill them with already defined values.
Currently, non-defined properties are removed.
This might be something we should not do.

File: org.eclipse.winery.frontends/app/tosca-management/src/app/instance/sharedComponents/properties/properties.component.ts [hunk 3]
```diff
@@ -25,55 +38,82 @@ import { HttpErrorResponse } from '@angular/common/http';
...
                 data => this.handleProperties(data),
-                error => this.handleError(error)
+                error => this.handleError(error, 'getProperties')
             );
     }
 
+    // TODO: handle inheritance
+    private getPropertiesDefinition() {
+        this._loading.getDefinitions = true;
+        // TODO: this does not send a request to the correct endpoint this endpoint url is constructed from browser url kekw
```
Comment (milesstoetzner) [inline org.eclipse.winery.frontends/app/tosca-management/src/app/instance/sharedComponents/properties/properties.component.ts:97]: discuss this

File: org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/entitytypes/properties/PropertiesDefinitionResource.java [hunk 4]
```diff
@@ -75,6 +78,59 @@ public PropertiesDefinitionResourceApiData getJson() {
...
+    // TODO: this merges only wineryPropertiesDefinitions?
+    // TODO: do we need to merge entityType#getProperties or something else?
+    // TODO: does not know at the end from which entityType the property definition has been derived from
+    @GET
+    @Path("merged")
+    @Produces(MediaType.APPLICATION_JSON)
+    public PropertiesDefinitionResourceApiData getMerged() {
+        ArrayList<TEntityType> parents = RepositoryFactory.getRepository().getParents(this.parentRes.getEntityType());
+        
+        // NOTE: this is not a deep copy but a reference!
```
Comment (lharzenetter) [inline org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/entitytypes/properties/PropertiesDefinitionResource.java:92]: If you want to have a dedicated list, you can use `new ArrayList(this.getEntityType()....);`
However, since this is only used to return the elements, this is okay. In case you would use it as an utils method, this is a valid point.

File: org.eclipse.winery.frontends/app/tosca-management/src/app/instance/sharedComponents/properties/properties.component.html [hunk 5]
```diff
@@ -11,33 +11,105 @@
...
+                                <p>The constraint "{{ form.controls[definition.key]?.errors?.constraint }}" is not satisfied!</p>
+                            </div>
+                            
+                        </ng-container>
+
+                        <p *ngIf="definition.type" class="text-muted">Type is "{{ definition.type }}".</p>
+                        <p *ngIf="!definition.type" class="text-muted">No type defined.</p>
+
+                        <p *ngIf="definition.defaultValue" class="text-muted">Default is "{{ definition.defaultValue }}
+                            ".</p>
```
Comment (lharzenetter) [inline org.eclipse.winery.frontends/app/tosca-management/src/app/instance/sharedComponents/properties/properties.component.html:85]: Why is this in a new line?

File: org.eclipse.winery.frontends/app/tosca-management/src/app/instance/sharedComponents/properties/properties.component.html [hunk 6]
```diff
@@ -11,33 +11,105 @@
...
+                </div>
+
+                <br/>
+                <br/>
+
+                <div class="form-group">
+
+                    <ng-container *ngFor="let definition of definitions; index as index">
+
+
```
Comment (lharzenetter) [inline org.eclipse.winery.frontends/app/tosca-management/src/app/instance/sharedComponents/properties/properties.component.html:49]: There are a lot of empty lines, I suggest to remove them, it makes it not more readable - on the contrary...

File: org.eclipse.winery.frontends/app/tosca-management/src/app/instance/sharedComponents/propertiesDefinition/propertiesDefinition.service.ts [hunk 7]
```diff
@@ -84,4 +101,8 @@ export class PropertiesDefinitionService {
     private sendJsonRequest<T>(requestPath: string): Observable<T> {
         return this.http.get<T>(backendBaseURL + this.route.url + requestPath);
     }
+
+
```
Comment (lharzenetter) [inline org.eclipse.winery.frontends/app/tosca-management/src/app/instance/sharedComponents/propertiesDefinition/propertiesDefinition.service.ts:105]: What's with these empty lines? Remove them

File: org.eclipse.winery.frontends/app/tosca-management/src/app/wineryTableModule/wineryTable.component.ts [hunk 8]
```diff
@@ -119,6 +119,8 @@ export class WineryTableComponent implements OnInit, DoCheck {
 
     @ViewChild('tableContainer') tableContainer: any;
     @Input() title: string;
+    @Input() titleRouterLink?: any[] | string;
```
Comment (lharzenetter) [inline org.eclipse.winery.frontends/app/tosca-management/src/app/wineryTableModule/wineryTable.component.ts:122]: Why is this an array of `any`?

File: org.eclipse.winery.frontends/app/tosca-management/src/app/wineryUtils/utils.ts [hunk 9]
```diff
@@ -21,6 +21,30 @@ export class Utils {
...
+    public static isLoading(map: {[key: string]: boolean}): boolean {
+        return Object.keys(map).some(k => map[k]);
+    }
+
+    public static doubleEncodeNamespace(namespace: string): string {
+        return encodeURIComponent(encodeURIComponent(namespace));
+    }
+
+    public static getFrontendPath(toscaType: string, namespace: string, localname: string) {
+        // TODO: this should be double encode?!
```
Comment (lharzenetter) [inline org.eclipse.winery.frontends/app/tosca-management/src/app/wineryUtils/utils.ts:33]: Normally it should, however, it may be the case that Angular is automatically encoding the already encoded elements a second time, thus it works in this case if its only encoded once. 
Does it work currently?

File: org.eclipse.winery.frontends/app/tosca-management/src/app/wineryValidators/wineryPropertyDefinitionValidator.directive.ts [hunk 10]
```diff
@@ -0,0 +1,71 @@
+/*******************************************************************************
+ * Copyright (c) 2017-2018 Contributors to the Eclipse Foundation
```
Comment (lharzenetter) [inline org.eclipse.winery.frontends/app/tosca-management/src/app/wineryValidators/wineryPropertyDefinitionValidator.directive.ts:2]: This seems to be a new file? Then the copyright header must be only `2022`

File: org.eclipse.winery.frontends/app/tosca-management/src/app/wineryValidators/wineryPropertyDefinitionValidator.module.ts [hunk 11]
```diff
@@ -0,0 +1,24 @@
+/*******************************************************************************
+ * Copyright (c) 2017 Contributors to the Eclipse Foundation
```
Comment (lharzenetter) [inline org.eclipse.winery.frontends/app/tosca-management/src/app/wineryValidators/wineryPropertyDefinitionValidator.module.ts:2]: See above.

**Review by lharzenetter — COMMENTED** [review]
If you changed a file, make sure you update also the license header to the current year.
Also, make sure that you appear in the NOTICE file.

File: org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/entitytypes/EntityTypeResource.java [hunk 12]
```diff
@@ -55,7 +55,7 @@ public TEntityType getEntityType() {
     public PropertiesDefinitionResource getPropertiesDefinitionResource() {
         return new PropertiesDefinitionResource(this);
     }
-
+
```
Comment (lharzenetter) [inline org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/entitytypes/EntityTypeResource.java:58]: Revert this change

File: org.eclipse.winery.frontends/app/tosca-management/src/app/wineryTableModule/wineryTable.component.ts (hunk 8 above)
Comment (milesstoetzner) [reply org.eclipse.winery.frontends/app/tosca-management/src/app/wineryTableModule/wineryTable.component.ts:122]: the definition is acutally taken from `RouterLinkWithHref`

```
export declare class RouterLinkWithHref implements OnChanges, OnDestroy {
    private router;
    private route;
    private locationStrategy;
    target: string;
    queryParams: {
        [k: string]: any;
    };
    fragment: string;
    queryParamsHandling: QueryParamsHandling;
    preserveFragment: boolean;
    skipLocationChange: boolean;
    replaceUrl: boolean;
    state?: {
        [k: string]: any;
    };
    private commands;
    private subscription;
    private preserve;
    href: string;
    constructor(router: Router, route: ActivatedRoute, locationStrategy: LocationStrategy);
    routerLink: any[] | string;
    preserveQueryParams: boolean;
    ngOnChanges(changes: {}): any;
    ngOnDestroy(): any;
    onClick(button: number, ctrlKey: boolean, metaKey: boolean, shiftKey: boolean): boolean;
    private updateTargetUrlAndHref;
    readonly urlTree: UrlTree;
}

```

Here is an example from the docs
> For instance ['/team', teamId, 'user', userName, {details: true}] means that we want to generate a link to /team/11/user/bob;details=true.


See `org.eclipse.winery.frontends/node_modules/@angular/router/src/directives/router_link.d.ts`

**Review by lharzenetter — APPROVED** [review]
LGTM! Please close PR and open a new one against the eclipse/main branch.

**Comment by milesstoetzner** [comment]
closed in favor of https://github.com/eclipse/winery/pull/660
